inc/elasticsearch_packages/namespace.php

```
@@ -0,0 +1,432 @@
+ */
+function packages_dir( string $path ) : string {
+	if ( ! defined( 'S3_UPLOADS_BUCKET' ) || empty( S3_UPLOADS_BUCKET ) ) {
+		return $path;
+	}
+
+	return sprintf( 's3://%s/es-packages', S3_UPLOADS_BUCKET );
+}
+
+/**
+ * Get Elasticsearch Service Client.
+ *
+ * @return ElasticsearchServiceClient
+ */
+function get_es_client() : ElasticsearchServiceClient {
+	return Altis\get_aws_sdk()->createElasticsearchService( [
+		'version' => '2015-01-01',
+	] );
+}
+
+/**
+ * Return the domain ID portion of the Elasticsearch Service host name.
+ */
+function get_elasticsearch_domain() : string {
+	return preg_replace( '#^search-([a-z][a-z0-9\-]+)-[a-z0-9]+\..*$#', '$1', ELASTICSEARCH_HOST );
+}
+
+/**
+ * Create the package on AES and return the package ID string.
+ *
+ * @param string|null $package_id The package file path for ES mappings.
+ * @param string $slug The package slug.
+ * @param string $file The package file path.
+ * @param bool $for_network Whether this is a network level package or not.
+ * @return string|WP_Error
+ */
+function create_package_id( ?string $package_id, string $slug, string $file, bool $for_network = false ) {
+
+	// Check file is an S3 file path.
+	if ( strpos( $file, 's3://' ) !== 0 ) {
+		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		return new WP_Error( 'package_file_missing', sprintf( 'The given package filepath %s is not a valid S3 file path.', $file ) );
+	}
+
+	// Check domain.
+	if ( ! defined( 'ELASTICSEARCH_HOST' ) || ! ELASTICSEARCH_HOST ) {
+		return new WP_Error( 'elasticsearch_host_not_found', 'Could not find an AWS ElasticSearch Service Domain to associate the package with.' );
+	}
+
+	// Get AES client.
+	$client = get_es_client();
+
+	// Get Domain ID from host.
+	$domain = get_elasticsearch_domain();
+
+	// Get a unique package name.
+	// Must be no more than 28 chars.
+	$name = substr( sprintf(
+		'p%s-%s',
+		hash( 'crc32', $domain . $file . time() ),
+		trim( basename( $file, '.txt' ), '-' )
```

Should this be sanitised?

---

Maybe, but AWS will throw an error if the string is invalid anyway.